---
title: "Ledesma_David_ADA_DATA_REANALYSIS_ASSIGNMENT"
author: "David Ledesma"
date: "4/4/2019"
output:
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

Gray et al. 2019 used geometric morphometrics to study morphological variation in the skulls of Australian agamid lizards. They were interested in testing whether diversity in skull shape can be attributed to heterochrony (change in the timing or rate of development) or is mainly a result of modifications to growth pathways. They tested whether ontogenetic trajectories are similar between species, if there is association between shape and phylogeny, and if growth patterns and life histories are correlated. 

I chose to replicate three analyses looking at 1) whether different life stages (adult and juvenile) have different amounts of shape disparity between species 2) whether any species exhibit isometric growth (growth without change in shape) 3) the relationship between growth, shape change, and life history. 

2D landmarks were placed on photographs of the skulls of 361 specimens spanning 18 species. 


## Visualization of Data
```{r}
library(curl)
library(geomorph)
f <- curl("https://raw.githubusercontent.com/TIMAVID/Ledesma_David_ADA_Data_Replication_Assignment/master/agamid.TPS")
raw_data <- readland.tps(f, specID = c("imageID")) # the function "readland.tps" reads the landmark data in tps format and returns a 3D array of the coordinate data
plot(raw_data)
head(raw_data)

f2 <- curl("https://raw.githubusercontent.com/TIMAVID/Ledesma_David_ADA_Data_Replication_Assignment/master/agamid.csv")
agamid <- read.csv(f2, header = TRUE, sep = ",", stringsAsFactors = TRUE) # this is a matrix of each specimen with associated data about the species, genus, habit, and clade  
head(agamid)
```

```{r}
new <- list(land = raw_data,species = agamid$species,habitat = agamid$habit,clade = agamid$clade, specimen = agamid$specimen) # combines corrdinate data, species, specimen, habit, and clade data
```
Before the cooridinate data can be analyzed, they must be alligned and size must be accounted for. This is done with a Generalized Procrustes Analysis.
```{r}
GPA_landmarks <- gpagen(new$land) # performs Generalized Procrustes analysis of landmarks and creates aligned Procrustes coordinates

new <- list(land = GPA_landmarks , species = agamid$species, habitat = agamid$habit, clade = agamid$clade, specimen = names(GPA_landmarks$Csize)) # the gpagen function changed some of names to lower case so new was remade with "names(GPA_landmarks$Csize"

plot(GPA_landmarks) # vizualize Procrustes alligned landmarks

gp <- as.factor(new$species) # creates factor defining species groups for PCA vizualization
PCA<-plotTangentSpace(GPA_landmarks$coords, axis1 = 1, axis2 = 2, groups = gp, legend = TRUE) #Pricipal component analysis + vizualization
PCA$pc.summary
```


![Figure 1 from Gray et al. 2019](/Users/davidledesma/Development/Ledesma_David_ADA_Data_Replication_Assignment/PCA.png )

A PCA serves as an exploratory data analysis of the morphospace of cranial shape variation. The points are color coded by species.  My replicated PCA plot is the same as the one published by Gray et al. 2019; however, mine seems to be flipped with respect to the PC2 axis. I'm not sure why this occurs and I was unable to flip my plot to match the published one. I  was also unable to figure out how to keep the legend from getting cut off in the plot. My PCA serves as adesprictive statistic by showing the variance explained by each PC axis which matches that reported in Gray et al. (2019). 

## Replications/Reanalysis

### Examining allometry ###
In order to quantify shape disparity between adults and juveniles in their sample, the authors split their sample into two groups: the smallest three specimens of each species and the largest three specimens of each species. The size of the specimens was determined by their centroid size derived from a Generalized Procrustes analysis. They used the fuction “morphol.disparity” in geomorph which calculates the Procrustes variance of each group (that is, calculating the variance using the Procrustes distances of each group). The variances are tested for significance by doing permutations that randomixe the rows of the shape matrix. 

```{r}
library(tidyverse)

# Data Processing

raw_data2 <- new 

# replace the "land" array (16 x 2 x 361 dimensions) in rawData with a list of 361 16 x 2 arrays

splitLand <- list(raw_data2)

for (i in 1:361){
  
  splitLand[[i]] <-raw_data[1:16,1:2,i]
  
}

raw_data2[["land"]] <- splitLand

# Specify a vector of the specimen names you want to extract

b <- data.frame(specimen = names(GPA_landmarks$Csize), size = (GPA_landmarks$Csize), species = (new$species)) # creates matrix combining specimen numbers, centroid sizes, and species

Adult_grp<- group_by(b, species) %>% # procedure for finding smallest three juveniles of each species and the largest three adults of each species based on centroid size
  mutate(rank = rank(-size)) %>%
  arrange(rank) %>% filter(rank <= 3)
Juv_grp<- group_by(b, species) %>%
  mutate(rank = rank(size)) %>%
  arrange(rank) %>% filter(rank <= 3)

ab<-Adult_grp[order(Adult_grp$specimen),]
ac<-Juv_grp[order(Juv_grp$specimen),]
ad<- rbind(ab, ac)
aa <- ad$specimen
aa <- as.vector(aa)


specimensToExtract <- c(aa)

# Create a tibble of just specimen names from your dataset and add a row ID column

new <- list(land = GPA_landmarks , species = agamid$species, habitat = agamid$habit, clade = agamid$clade, specimens = names(GPA_landmarks$Csize))

specimens <- tibble(specimenName=as.character(new$specimen))

specimens <- rowid_to_column(specimens, "ID")



# Filter that tibble to include only the ones you want to extract

specimens <- filter(specimens,specimenName %in% specimensToExtract)


# Set up vectors of variables to pull out for each specimen

land <- vector("list",nrow(specimens))

species <- vector()

habitat <- vector()

clade <- vector()

specimen <- vector()



# Extract from rawData just the specimens for which you want data

j <- 0

for (i in specimens$ID){
  
  j <- j + 1
  
  land[[j]] <- raw_data2[["land"]][[i]]
  
  species <- c(species,as.character(raw_data2[["species"]][[i]]))
  
  habitat <- c(habitat,as.character(raw_data2[["habitat"]][[i]]))
  
  clade <- c(clade,as.character(raw_data2[["clade"]][[i]]))
  
  specimen <- c(specimen,as.character(raw_data2[["specimen"]][[i]]))
  
}


# Assemble the extracted data into a format like that of your original dataset

raw_data2 <- list("land"=unlist(land),"species"=as.factor(species),"habitat"=as.factor(habitat),"clade"=as.factor(clade),"specimen"=as.factor(specimen)) #raw data of the smallest three juveniles of each species and the largest three adults of each species



# the lines below recast the list in "land" to the original array format and attributes

dim(raw_data2$land) <- c(16,2,nrow(specimens))

attributes(raw_data2$land)$dimnames[[3]] <- specimens$specimenName

```

```{r}
plot(raw_data2$land) 
```

```{r}
n.rows <- 54 # procedure for creating a matrix that will be used for defining age classes
n.cols <- 1
n.e <- n.rows * n.cols
my.matrix.2 <- matrix(rep(NA, n.e), ncol = n.cols)
for (i in 1:n.cols){
  for (j in 1:n.rows){
    my.matrix.2[j,i] <- paste("adult")
  }
}
my.matrix.3 <- matrix(rep(NA, n.e), ncol = n.cols)
for (i in 1:n.cols){
  for (j in 1:n.rows){
    my.matrix.3[j,i] <- paste("juvenile")
  }
}

my.matrix <- rbind(my.matrix.2, my.matrix.3) # create maxtrix of age classes

adult<- group_by(Adult_grp, species)
adult<- (adult$specimen)
adult<- as.character(adult)

juvenile <- group_by(Juv_grp, species)
juvenile<- (juvenile$specimen)
juvenile<- as.character(juvenile)

az <- c(adult, juvenile)

wow<- list(specimen = az, age = as.factor(my.matrix))
wow <- as.data.frame(wow)

specimen <- as.data.frame(specimen)
species_grps <- inner_join(specimen, wow)

```
```{r}
GPA_ontogeny <- gpagen(raw_data2$land) # performs Generalized Procrustes analysis of landmarks

gdf_ontogeny <- geomorph.data.frame(GPA_ontogeny, Age = species_grps$age) # creates Geomorph data frame combinging GPA alligned landmarks with age classes

morphol.disparity(coords ~ Csize, groups= ~Age, data = gdf_ontogeny, iter=999, print.progress = FALSE) # calculates Procrustes variance between adults and juveniles while accounting for group size. Uses absolute differences in variances to test for pairwise differences between groups. The statistical significance of the differences between groups is assessed using a randomised residual permutation test with "iter= #" specified
age_grps <- as.factor(my.matrix) # creates factor defining age groups to be used for PCA
PCA2<- plotTangentSpace(gdf_ontogeny$coords, groups = species_grps$age, legend = TRUE) # PCA
```

### Test for isometric growth ###
In order to test whether any species exhibited isometric growth (growth without a change in shape), the authors performed a linear model regression of the log transformed centroid size (Csize) onto the shape for each species. Statistical significance was assesed through permutations with randomization of the residuals. 

```{r}
gdf <- geomorph.data.frame(GPA_landmarks, species = new$species, habitat = new$habitat, clade = new$clade, specimen= new$specimen) # creates Geomorph data frame of landmarks combined with additional data

#makes new coordinate groups according to species
group <- factor(paste(gdf$species))
levels(group)
new.coords_iso <- coords.subset(A = GPA_landmarks$coords, group = group)
names(new.coords_iso) # see the list levels

```

```{r}
#test for the association between centroid size and shape for each species
barbata_landmarks <- gpagen(new.coords_iso$barbata,  print.progress = FALSE)
barbata_df <- geomorph.data.frame(barbata_landmarks)
barbata_iso <- procD.lm(two.d.array(coords) ~ log(Csize), data = barbata_df, iter = 999,  print.progress = FALSE) # two.d.array function converts a 3D array to a 2D array

caudicinctus_landmarks <- gpagen(new.coords_iso$caudicinctus,  print.progress = FALSE)
caudicinctus_df <- geomorph.data.frame(caudicinctus_landmarks)
caudicinctus_iso <- procD.lm(two.d.array(coords) ~ log(Csize), data = caudicinctus_df, iter = 999,  print.progress = FALSE)

cristatus_landmarks <- gpagen(new.coords_iso$cristatus,  print.progress = FALSE)
cristatus_df <- geomorph.data.frame(cristatus_landmarks)
cristatus_ios <- procD.lm(two.d.array(coords) ~ log(Csize), data = cristatus_df, iter = 999,  print.progress = FALSE)

decresii_landmarks <- gpagen(new.coords_iso$decresii,  print.progress = FALSE)
decresii_df <- geomorph.data.frame(decresii_landmarks)
decresii_iso <- procD.lm(two.d.array(coords) ~ log(Csize), data = decresii_df, iter = 999,  print.progress = FALSE)

diemensis_landmarks <- gpagen(new.coords_iso$diemensis,  print.progress = FALSE)
diemensis_df <- geomorph.data.frame(diemensis_landmarks)
diemensis_iso <- procD.lm(two.d.array(coords) ~ log(Csize), data = diemensis_df, iter = 999,  print.progress = FALSE)

gilberti_landmarks <- gpagen(new.coords_iso$gilberti,  print.progress = FALSE)
gilberti_df <- geomorph.data.frame(gilberti_landmarks)
gilberti_iso <- procD.lm(two.d.array(coords) ~ log(Csize), data = gilberti_df, iter = 999,  print.progress = FALSE)

horridus_landmarks <- gpagen(new.coords_iso$horridus,  print.progress = FALSE)
horridus_df <- geomorph.data.frame(horridus_landmarks)
horridus_iso <- procD.lm(two.d.array(coords) ~ log(Csize), data = horridus_df, iter = 999,  print.progress = FALSE)

isolepis_landmarks <- gpagen(new.coords_iso$isolepis,  print.progress = FALSE)
isolepis_df <- geomorph.data.frame(isolepis_landmarks)
isolepis_iso <- procD.lm(two.d.array(coords) ~ log(Csize), data = isolepis_df, iter = 999,  print.progress = FALSE)

kingii_landmarks <- gpagen(new.coords_iso$kingii,  print.progress = FALSE)
kingii_df <- geomorph.data.frame(kingii_landmarks)
kingii_iso <- procD.lm(two.d.array(coords) ~ log(Csize), data = kingii_df, iter = 999,  print.progress = FALSE)

lesueurii_landmarks <- gpagen(new.coords_iso$lesueurii,  print.progress = FALSE)
lesueurii_df <- geomorph.data.frame(lesueurii_landmarks)
lesueurii_iso <- procD.lm(two.d.array(coords) ~ log(Csize), data = lesueurii_df, iter = 999,  print.progress = FALSE)

longirostris_landmarks <- gpagen(new.coords_iso$longirostris,  print.progress = FALSE)
longirostris_df <- geomorph.data.frame(longirostris_landmarks)
longirostris_iso <- procD.lm(two.d.array(coords) ~ log(Csize), data = longirostris_df, iter = 999,  print.progress = FALSE)

muricatus_landmarks <- gpagen(new.coords_iso$muricatus,  print.progress = FALSE)
muricatus_df <- geomorph.data.frame(muricatus_landmarks)
muricatus_iso <- procD.lm(two.d.array(coords) ~ log(Csize), data = muricatus_df, iter = 999,  print.progress = FALSE)

nobbi_landmarks <- gpagen(new.coords_iso$nobbi,  print.progress = FALSE)
nobbi_df <- geomorph.data.frame(nobbi_landmarks)
nobbi_iso <- procD.lm(two.d.array(coords) ~ log(Csize), data = nobbi_df, iter = 999,  print.progress = FALSE)

nuchalis_landmarks <- gpagen(new.coords_iso$nuchalis,  print.progress = FALSE)
nuchalis_df <- geomorph.data.frame(nuchalis_landmarks)
nuchalis_iso <- procD.lm(two.d.array(coords) ~ log(Csize), data = nuchalis_df, iter = 999,  print.progress = FALSE)

reticulatus_landmarks <- gpagen(new.coords_iso$reticulatus,  print.progress = FALSE)
reticulatus_df <- geomorph.data.frame(reticulatus_landmarks)
reticulatus_iso <- procD.lm(two.d.array(coords) ~ log(Csize), data = reticulatus_df, iter = 999,  print.progress = FALSE)

tetraporophora_landmarks <- gpagen(new.coords_iso$tetraporophora,  print.progress = FALSE)
tetraporophora_df <- geomorph.data.frame(tetraporophora_landmarks)
tetraporophora_iso <- procD.lm(two.d.array(coords) ~ log(Csize), data = tetraporophora_df, iter = 999,  print.progress = FALSE)

vitticeps_landmarks <- gpagen(new.coords_iso$vitticeps,  print.progress = FALSE)
vitticeps_df <- geomorph.data.frame(vitticeps_landmarks)
vitticeps_iso <- procD.lm(two.d.array(coords) ~ log(Csize), data = vitticeps_df, iter = 999,  print.progress = FALSE)

winneckei_landmarks <- gpagen(new.coords_iso$winneckei,  print.progress = FALSE)
winneckei_df <- geomorph.data.frame(winneckei_landmarks)
winneckei_iso <- procD.lm(two.d.array(coords) ~ log(Csize), data = winneckei_df, iter = 999,  print.progress = FALSE)
```

```{r}
iso <- rbind(barbata_iso$aov.table[1,7], caudicinctus_iso$aov.table[1,7], cristatus_ios$aov.table[1,7], decresii_iso$aov.table[1,7], diemensis_iso$aov.table[1,7], 
             gilberti_iso$aov.table[1,7], horridus_iso$aov.table[1,7], isolepis_iso$aov.table[1,7], kingii_iso$aov.table[1,7], lesueurii_iso$aov.table[1,7], longirostris_iso$aov.table[1,7], 
             muricatus_iso$aov.table[1,7], nobbi_iso$aov.table[1,7],nuchalis_iso$aov.table[1,7], reticulatus_iso$aov.table[1,7], tetraporophora_iso$aov.table[1,7], vitticeps_iso$aov.table[1,7], winneckei_iso$aov.table[1,7])
iso <- cbind(names(new.coords_iso), iso)
colnames(iso) <- c("species", "P-value")
iso
```

### Testing for ontogenetic allometric patterns among life habit groups ###
In order to test whether different life habit types had significant differences in terms of ontogenetic allometric patterns, the authors performed a multivariate analysis of covariance. They did this with the function "procD.allometry" and specified a model that includes log transformed centroid size, life habit, and their interaction. Statistical significance was tested with randomised residual permutations and Goodall’s F-ratio test. 
```{r}
# The origional data "agamid" had Arboreal as one life habits; however, in the analyses they were treated as Semi-arboreal. The following code changes all Arboreal classifications to Semi-arboreal and creates a new data frame with this data.
habit <- agamid$habit
levels(habit) <- gsub("Arboreal", "Semi-arboreal", levels(habit))
gdf2 <- geomorph.data.frame(GPA_landmarks, species = new$species, habitat = habit, clade = new$clade, specimen= new$specimen) # creates Geomorph data frame of landmarks combined with additional data
```
```{r}
# the following code uses the "procD.lm" function to perform a MANCOVA with size, habitat, and their interaction as parameters 
agamid_Allometry1 <-procD.lm(two.d.array(coords) ~ log(Csize) * habitat, data = gdf2, iter = 999, print.progress = FALSE) #
summary(agamid_Allometry1)
```


![Table S3 from Gray et al. 2019](/Users/davidledesma/Development/Ledesma_David_ADA_Data_Replication_Assignment/Table1.png )

```{r}
# the following code uses the function "advanced.procD.lm" which performes a Procrustes ANOVA and calculates whether differences are significant through pairwise tests and permutations. 
habit_allom <- advanced.procD.lm(f1= coords ~ log(Csize),
                                 f2= ~NULL, groups = ~ habitat,
                                 slope = ~ log(Csize), angle.type = "deg", iter = 999, data = gdf2, pc.shape = TRUE, print.progress = FALSE)



```
```{r}
data.frame(P_Angle = habit_allom$P.angles) 
data.frame(P_Length = habit_allom$P.slopes.dist)
```


![Table S4 from Gray et al. 2019](/Users/davidledesma/Development/Ledesma_David_ADA_Data_Replication_Assignment/Table2.png)




## Summar

## References

Gray, J. A., E. Sherratt, M. N. Hutchinson, & M. E. Jones. 2019. Changes in ontogenetic patterns facilitate diversification in skull shape of Australian agamid lizards. BMC Evolutionary Biology 19:7. 
